develop/tutorials/articles/260-themes-and-layout-templates/01-themes/overwriting-gulp-theme-tasks.markdown

```diff
@@ -0,0 +1,113 @@
+
+1.  Identify the gulp task or sub task that you want to hook into or overwrite. 
+		The tasks and their sub tasks are listed in their `[task-name].js` file in 
+		the 
+		[`tasks` folder](https://github.com/liferay/liferay-theme-tasks/tree/master/tasks)
+		of the 
+		[`liferay-theme-tasks`](https://github.com/liferay/liferay-theme-tasks)
+		repo. For example, the gulp `build` task and sub tasks are defined in the 
+		[`build.js` file](https://github.com/liferay/liferay-theme-tasks/blob/master/tasks/build.js#L38-L53): 
+
+				'build:clean',
+				'build:base',
+				'build:src',
+				'build:web-inf',
+				'build:liferay-look-and-feel',
+				'build:hook',
+				'build:themelets',
+				'build:rename-css-dir',
+				'build:prep-css',
+				'build:compile-css',
+				'build:fix-url-functions',
+				'build:move-compiled-css',
+				'build:remove-old-css-dir',
+				'build:fix-at-directives',
+				'build:r2',
+				'build:war',
+
+2.  Open your theme's `gulpfile.js` file and locate the 
+		`liferayThemeTasks.registerTasks()` method. This method registers the 
+		default gulp theme tasks. Add the `hookFn` property to the `registerTasks()` 
+		method's configuration object, making sure to pass in the `gulp` instance:
+
+				liferayThemeTasks.registerTasks({
+					gulp: gulp,
+					hookFn: function(gulp) {
+						
+					}
+				});
+		
+3.  Inside the `hookFn()` function, use the `gulp.hook()` method to specify the 
+		theme task or sub task that you want to hook into. You can inject your code 
+		before or after a task by prefixing the `before:` or `after:` keyword before 
+		the task. Alternatively, you can use the `gulp.task()` method to overwrite a 
+		gulp task. Both methods have two parameters: the task or sub task you want 
+		to hook into and a callback function, which invokes `done` or returns a 
+		stream, with the logic that you want to inject. A few example configuration 
+		patterns are shown below:
+		
+				liferayThemeTasks.registerTasks({
+					gulp: gulp,
+					hookFn: function(gulp) {
+						gulp.hook('before:build:src', function(done) {
+							// Fires before build:src task
+						});
+
+						gulp.hook('after:build', function(done) {
+							// Fires after build task
+						});
+
+						gulp.task('build:base', function(done) {
+							// Overwrites build:base task
+						});
+					}
+				});	 
+
+The example below fires before the `build:src` sub-task and reads the JavaScript 
+files in the theme's `src` folder, minifies them with the `gulp-uglify` module, 
+places it in the `./src/js` folder, and invokes `done`:
+
+		'use strict';
+
+		var gulp = require('gulp');
+		var uglify = require('gulp-uglify');
+		var liferayThemeTasks = require('liferay-theme-tasks');
+
+		liferayThemeTasks.registerTasks({
+			gulp: gulp,
+			hookFn: function(gulp) {
+				gulp.hook('before:build:src', function(done) {
+					// Fires after build `src` task
+					gulp.src('./src/js/*.js')
+					.pipe(uglify())
+					.pipe(gulp.dest('./src/js'))
```

Comment:
If someone does this, it will actually overwrite their source files. Can we provide a different example?

Reply:
Of course. This is just one I tested myself. Can you provide me with one? or a small change to this one that will make it follow best practices.

Reply:
@Robert-Frampton the idea is that this would minify your JS code before the war file is built. When should this run if not before the `build:src` task? If I don't overwrite the `src` JS files with this task, how will the war file get the changes? Thanks!

Reply:
You could create a hook that fires after the files have been moved to the build directory, but before the war has been generated. Then in the hook you would grab the files in build, minify them, and then place them back into the build directory.

Reply:
K. I'll update the example to do that. Thanks!

Reply:
@Robert-Frampton Here's the updated example.
```
	hookFn: function(gulp) {
		gulp.hook('before:build:war', function(done) {
	        // Fires before build war task
		gulp.src('./build/js/*.js')
		.pipe(uglify())
		.pipe(gulp.dest('./build/js'))
		.on('end', done);
		});
	}
```